cmd/dashboard/main.go

```diff
@@ -0,0 +1,134 @@
+	var (
+		m      = metrics.NewMetrics("dashboard")
+		logger = log.New("app", "Dashboard")
+		ctx    = context.Background()
+	)
+
+	l1, err := eth.Dial(rpcURL, l1Contract, logger)
+	if err != nil {
+		log.Crit("Failed to create L1 source", "err", err)
+	}
+
+	shardEntryBits, err := readUintFromContract(ctx, l1.Client, l1Contract, "shardEntryBits")
+	if err != nil {
+		return nil, err
+	}
+
+	return &dashboard{
+		ctx:        ctx,
+		l1Source:   l1,
+		m:          m,
+		l1Contract: l1Contract,
+		kvEntries:  1 << shardEntryBits,
+		logger:     logger,
+	}, nil
+}
+
+func (d *dashboard) RefreshContractMetrics(ctx context.Context, sig eth.L1BlockRef) {
+	lastKVIndex, err := d.l1Source.GetStorageLastBlobIdx(int64(sig.Number))
+	if err != nil {
+		log.Warn("Refresh contract metrics (last kv index) failed", "err", err.Error())
+		return
+	}
+	maxShardIdx := lastKVIndex / d.kvEntries
+	d.m.SetLastKVIndexAndMaxShardId(sig.Number, lastKVIndex, maxShardIdx)
+	d.logger.Info("RefreshContractMetrics", "lastKvIndex", lastKVIndex, "maxShardIdx", maxShardIdx)
+
+	l1api := miner.NewL1MiningAPI(d.l1Source, d.logger)
+	for sid := uint64(0); sid <= maxShardIdx; sid++ {
+		info, err := l1api.GetMiningInfo(ctx, d.l1Contract, sid)
+		if err != nil {
+			log.Warn("Get mining info for metrics fail", "shardId", sid, "err", err.Error())
+			continue
+		}
+		d.m.SetMiningInfo(sid, info.Difficulty.Uint64(), info.LastMineTime, info.BlockMined.Uint64())
+		d.logger.Info("Refresh mining info", "blockMined", info.BlockMined, "lastMineTime", info.LastMineTime, "Difficulty", info.Difficulty)
+	}
+}
+
+func readSlotFromContract(ctx context.Context, client *ethclient.Client, l1Contract common.Address, fieldName string) ([]byte, error) {
+	h := crypto.Keccak256Hash([]byte(fieldName + "()"))
+	msg := ethereum.CallMsg{
+		To:   &l1Contract,
+		Data: h[0:4],
+	}
+	bs, err := client.CallContract(ctx, msg, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get %s from contract: %v", fieldName, err)
+	}
+	return bs, nil
+}
+
+func readUintFromContract(ctx context.Context, client *ethclient.Client, l1Contract common.Address, fieldName string) (uint64, error) {
+	bs, err := readSlotFromContract(ctx, client, l1Contract, fieldName)
+	if err != nil {
+		return 0, err
+	}
+	value := new(big.Int).SetBytes(bs).Uint64()
+	log.Info("Read uint from contract", "field", fieldName, "value", value)
+	return value, nil
+}
+
+func main() {
+	// Parse the flags and set up the logger to print everything requested
+	flag.Parse()
+	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(*logFlag), log.StreamHandler(os.Stderr, log.TerminalFormat(true))))
+
+	if *portFlag < 0 || *portFlag > math.MaxUint16 {
+		log.Crit("Invalid port")
+	}
+	d, err := newDashboard(*rpcURLFlag, common.HexToAddress(*l1ContractFlag))
+	if err != nil {
+		log.Crit("New dashboard fail", "err", err)
+	}
+	l1LatestBlockSub := eth.PollBlockChanges(d.ctx, d.logger, d.l1Source, d.RefreshContractMetrics, ethRPC.LatestBlockNumber,
```

Comment:
If the dashboard crashes or stops for some time, we may need to start monitoring the data from where it last stopped